### PR TITLE
Add U-Boot 2024.04

### PR DESCRIPTION
**Pull request by janaperic**
U-Boot 2020.01 will remain default options for the time being, but we want version 2024.04 to be available.

### CODE REVIEW

**Comment by lluiscampos** [comment]
@TheYoctoJester I don't know how to review this PR without doing the work myself in parallel. Can you elaborate a bit on which patches applied cleanly and which parts did you need to modify? :pray:

**Comment by TheYoctoJester** [comment]
@lluiscampos @janaperic sorry it took a bit.

So the first five patches (#[8b806f3](https://github.com/mendersoftware/uboot-mender/pull/9/commits/8b806f3114fef94674353712f2f86bb7cd3bbf03), #[38938d2](https://github.com/mendersoftware/uboot-mender/pull/9/commits/38938d29e8c20afb2f637b4a932f9f84b1db3bc7), #[90c2046](https://github.com/mendersoftware/uboot-mender/pull/9/commits/90c2046a1fa344aadc96412616c680a2fe3f525d), #[450288f](https://github.com/mendersoftware/uboot-mender/pull/9/commits/450288f5c9c0f6fcb13d050d18be26470b438d25) and #[99a4791](https://github.com/mendersoftware/uboot-mender/pull/9/commits/99a47914aefc3bea638d6099f604e9ba90051b50)) are directly taken from the [`meta-mender`](https://github.com/mendersoftware/meta-mender/tree/master/meta-mender-core/recipes-bsp/u-boot/patches),  [`meta-mender-community/meta-mender-raspberrypi`](https://github.com/mendersoftware/meta-mender-community/tree/scarthgap/meta-mender-raspberrypi/recipes-bsp/u-boot/patches) and the [`mender-rpi-2020.01` branch](https://github.com/mendersoftware/uboot-mender/commit/4cc786ae5c2706afa44487ad1a1c886b8d4d89a1). There are no functional modifications taken.

The [last commit](https://github.com/mendersoftware/uboot-mender/pull/9/commits/c4fba9bb858da75907bcad4fab2153ff23a20437) adds a special case for the Raspberry Pi family in `BOOTCMD`. In earlier versions, the file `boot.scr` was used to inject the Raspberry Pi specific boot process, which is
- do not load DTB, but use the (patched) one provided by the first stage loader
- extract `cmdline` from the first stage loader provided one, and extend
This process of loading that from `boot.scr` is not supported by `u-boot` anymore, so the logic has been hardcoded into the resulting `u-boot` environment.

File: include/env_mender.h [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+
+#if defined(CONFIG_TARGET_RPI) || \
+    defined(CONFIG_TARGET_RPI_2) || \
+    defined(CONFIG_TARGET_RPI_0_W) || \
+    defined(CONFIG_TARGET_RPI_3) || \
+    defined(CONFIG_TARGET_RPI_3_32B) || \
+    defined(CONFIG_TARGET_RPI_4) || \
+    defined(CONFIG_TARGET_RPI_4_32B) || \
+    defined(CONFIG_TARGET_RPI_ARM64)
+#define CONFIG_MENDER_BOOTCOMMAND                                       \
```
Comment (lluiscampos) [inline include/env_mender.h]: For consistency across this file:
```suggestion
# define CONFIG_MENDER_BOOTCOMMAND                                       \
```

(same below after the `#else`, if this suggestion is to be accepted)

File: include/env_mender.h [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+    "fdt addr ${fdt_addr} && fdt get value bootargs /chosen bootargs; " \
+    "run mender_setup; "                                                \
+    "mmc dev ${mender_uboot_dev}; "                                     \
+    "if load ${mender_uboot_root} ${kernel_addr_r} /boot/vmlinux; then " \
+    "booti ${kernel_addr_r} - ${fdt_addr}; "                            \
+    "fi; "                                                              \
+	"echo \"could not load kernel, trying to recover\"; "               \
```
Comment (lluiscampos) [inline include/env_mender.h]: Mix of spaces and tabs

**Review by lluiscampos — APPROVED** [review]
> @lluiscampos @janaperic sorry it took a bit. (...)

No problem. Thank you very much.

It looks good. I added two nit-pick suggestions for the last commit.

File: include/configs/rpi.h [hunk 3]
```diff
@@ -31,4 +31,9 @@
  */
 #define CFG_SYS_SDRAM_SIZE		SZ_128M
 
+/* Environment */
+ 
+#define CONFIG_BOOTCOUNT_ENV
```
Comment (TheMeaningfulEngineer) [inline include/configs/rpi.h]: (Mindprobe: in my mind these setting are doing the same thing as modifying the .config, they are just set in the header file.)

You're already modifying the defconfigs.
Why have this separately set and in a header file?

Or if the reason was to not repeat the same config for every defconfig, why not translate
```
CONFIG_ENV_IS_IN_MMC=y
CONFIG_ENV_OFFSET=0x400000
CONFIG_ENV_OFFSET_REDUND=0x1000000
CONFIG_SYS_REDUNDAND_ENVIRONMENT=y
```

here and remove all the defconfig patches?

Comment (TheYoctoJester) [reply include/configs/rpi.h]: @TheMeaningfulEngineer after thinking a bit about it, actually I think the header modifications should be removed. It seems to be historic clutter for unknown reasons, but I'd definitely say that those settings should go to the respective `defconfigs`, just like they do on Yocto already. Hence moved them there to keep things in line.

**Review by lluiscampos — APPROVED** [review]

**Pull request merged**